KeyVault.Acmebot/Providers/NamecheapProvider.cs

```diff
@@ -0,0 +1,561 @@
+
+    private NamecheapGetHostsResponse ParseGetHostsResponse(string responseBody)
+    {
+        var xmlDoc = new XmlDocument() { PreserveWhitespace = true };
+        xmlDoc.LoadXml(responseBody);
+        var nsMgr = CreateNamespaceManager(xmlDoc.NameTable);
+        var getHostsResponse = new NamecheapGetHostsResponse();
+
+        // Parse DomainDNSGetHostsResult element
+        var getHostsResultElement = xmlDoc.SelectSingleNode(@"/nc:ApiResponse/nc:CommandResponse/nc:DomainDNSGetHostsResult", nsMgr);
+        getHostsResponse.EmailType = getHostsResultElement.Attributes["EmailType"]?.Value;
+        getHostsResponse.Domain = getHostsResultElement.Attributes["Domain"]?.Value;
+
+        // Parse host elements
+        var hostElements = getHostsResultElement.SelectNodes(@"nc:host", nsMgr);
+
+        foreach (XmlNode hostElement in hostElements)
+        {
+            getHostsResponse.Hosts.Add(new NamecheapHost
+            {
+                HostName = hostElement.Attributes["Name"]?.Value,
+                RecordType = hostElement.Attributes["Type"]?.Value,
+                Address = hostElement.Attributes["Address"]?.Value,
+                MxPref = hostElement.Attributes["MXPref"]?.Value,
+                Ttl = hostElement.Attributes["TTL"]?.Value
+            });
+        }
+
+        // Parse Status and Errors
+        ParseBaseElements(getHostsResponse, xmlDoc, nsMgr);
+
+        return getHostsResponse;
+    }
+
+    private async Task<string> CreateRelativeUriAsync(string command, DnsZone domain, string clientIp = null)
+    {
+        string relativeUri = await CreateRelativeUriCommonAsync(command, clientIp);
+
+        if (domain != null)
+        {
+            int dot = domain.Name.IndexOf('.');
+            string sld = domain.Name.Substring(0, dot);
+            string tld = domain.Name.Substring(dot + 1);
+            relativeUri += "&SLD=" + sld;
+            relativeUri += "&TLD=" + tld;
+        }
+
+        return relativeUri;
+    }
+
+    private async Task<string> CreateRelativeUriAsync(string command, int? page = null, string clientIp = null)
+    {
+        string relativeUri = await CreateRelativeUriCommonAsync(command, clientIp);
+
+        if (page != null)
+        {
+            relativeUri += $"&Page={page}";
+        }
+
+        return relativeUri;
+    }
+
+    private async Task<string> CreateRelativeUriCommonAsync(string command, string clientIp = null)
+    {
+        if (string.IsNullOrEmpty(clientIp))
+        {
+            clientIp = await ResolvePublicIPAsync();
+        }
+
+        return $"?ApiUser={ApiUser}&ApiKey={ApiKey}&UserName={Username}&ClientIp={clientIp}&Command={command}";
+    }
+
+    private async Task<string> ResolvePublicIPAsync()
+    {
+        var response = await _ipfyOrgHttpClient.GetAsync("?format=json");
+
+        response.EnsureSuccessStatusCode();
+
+        var ipFyResponse = await response.Content.ReadAsAsync<IpfyResponse>();
+
+        return ipFyResponse.Ip;
```

Comment:
Azure Functions do not always match the Outbound IP between each request. Therefore, there is a possibility of unintended behavior when the obtained Outbound IP changes.